llvm/lib/SYCLLowerIR/LowerESIMD.cpp

```diff
@@ -1272,68 +1286,102 @@ size_t SYCLLowerESIMDPass::runOnFunction(Function &F,
+        translatePackMask(*CI);
         ESIMDToErases.push_back(CI);
         continue;
       }
-    }
+      if (Name.startswith("__esimd_unpack_mask")) {
+        translateUnPackMask(*CI);
+        ESIMDToErases.push_back(CI);
+        continue;
+      }
+      // If vload/vstore is not about the vector-types used by
+      // those globals marked as genx_volatile, We can translate
+      // them directly into generic load/store inst. In this way
+      // those insts can be optimized by llvm ASAP.
+      if (Name.startswith("__esimd_vload")) {
+        if (translateVLoad(*CI, GVTS)) {
+          ESIMDToErases.push_back(CI);
+          continue;
+        }
+      }
+      if (Name.startswith("__esimd_vstore")) {
+        if (translateVStore(*CI, GVTS)) {
+          ESIMDToErases.push_back(CI);
+          continue;
+        }
+      }
 
-    if (Name.startswith("__esimd_get_value")) {
-      translateGetValue(*CI);
-      ESIMDToErases.push_back(CI);
-      continue;
-    }
+      if (Name.startswith("__esimd_get_value")) {
+        translateGetValue(*CI);
+        ESIMDToErases.push_back(CI);
+        continue;
+      }
 
-    if (Name.consume_front(SPIRV_INTRIN_PREF)) {
-      translateSpirvIntrinsic(CI, Name, ESIMDToErases);
-      // For now: if no match, just let it go untranslated.
-      continue;
+      if (Name.empty() || !Name.startswith(ESIMD_INTRIN_PREF1))
+        continue;
+      // this is ESIMD intrinsic - record for later translation
+      ESIMDIntrCalls.push_back(CI);
     }
 
-    if (Name.empty() || !Name.startswith(ESIMD_INTRIN_PREF1))
-      continue;
-    // this is ESIMD intrinsic - record for later translation
-    ESIMDIntrCalls.push_back(CI);
+    // Translate loads from SPIRV builtin globals into GenX intrinsics
+    auto *LI = dyn_cast<LoadInst>(&I);
+    if (LI) {
+      Value *LoadPtrOp = LI->getPointerOperand();
+      Value *SpirvGlobal = nullptr;
+      // Look through casts to find SPIRV builtin globals
+      auto *CE = dyn_cast<ConstantExpr>(LoadPtrOp);
+      if (CE) {
+        assert(CE->isCast() && "ConstExpr should be a cast");
+        SpirvGlobal = CE->getOperand(0);
+      } else {
+        SpirvGlobal = LoadPtrOp;
+      }
+
+      if (isa<GlobalVariable>(SpirvGlobal) &&
+          SpirvGlobal->getName().startswith(SPIRV_INTRIN_PREF)) {
+        auto PrefLen = StringRef(SPIRV_INTRIN_PREF).size();
+
+        // Go through all the uses of the load instruction from SPIRV builtin
+        // globals, which are required to be extractelement instructions.
+        // Translate each of them.
+        for (auto *LU : LI->users()) {
+          auto *EEI = dyn_cast<ExtractElementInst>(LU);
+          assert(EEI && "User of load from global SPIRV builtin is not an "
+                        "extractelement instruction");
+          Value *TranslatedVal = translateSpirvIntrinsic(
+              EEI, SpirvGlobal->getName().drop_front(PrefLen));
+
+          if (TranslatedVal) {
```

Comment:
should this be rather an assert(TranslatedVal)? Failure to translate a load from a SPIR-V intrinsic var seems like an error.

Reply:
Yes. Will change.